source/common/logger/ur_sinks.hpp

```diff
@@ -48,26 +49,41 @@ class Sink {
         flush_level = logger::Level::ERR;
     }
 
+    void flush_buffer() {
+        std::scoped_lock<std::mutex> output_lock(output_mutex);
+        get_buffer().sync();
+        get_buffer().flush();
+        *ostream << get_buffer().rdbuf();
+        ostream->flush();
+        get_buffer().str("");
+    }
+
   private:
     std::string logger_name;
     bool skip_prefix;
+    std::mutex output_mutex;
+
+    std::stringstream &get_buffer() {
+        static thread_local std::stringstream buffer;
```

Comment:
If this is inside the Sink class, then I think all instances of Sink will share it (including Stdout and File Sinks)- is this what you want?

Reply:
Moved buffer implementation to derived classes, so that every Sink type has unique buffer now. However, still the buffer can be shared between logger instances with the same sink types.